lib/internal/streams/readable.js

```diff
@@ -1382,29 +1488,31 @@ function fromList(n, state) {
     state.buffer.clear();
   } else {
     // read part of list.
-    ret = state.buffer.consume(n, state.decoder);
+    ret = state.buffer.consume(n, decoder);
   }
 
   return ret;
 }
 
 function endReadable(stream) {
   const state = stream._readableState;
+  const endEmitted = (state.state & kEndEmitted) !== 0;
 
-  debug('endReadable', state.endEmitted);
-  if (!state.endEmitted) {
-    state.ended = true;
+  debug('endReadable', endEmitted);
+  if (!endEmitted) {
+    state.state |= kEnded;
     process.nextTick(endReadableNT, state, stream);
   }
 }
 
 function endReadableNT(state, stream) {
-  debug('endReadableNT', state.endEmitted, state.length);
+  const endEmitted = (state.state & kEndEmitted) !== 0;
+
+  debug('endReadableNT', endEmitted, state.length);
 
   // Check that we didn't get one last unshift.
-  if (!state.errored && !state.closeEmitted &&
-      !state.endEmitted && state.length === 0) {
-    state.endEmitted = true;
+  if ((state.state & (kErrored | kCloseEmitted | kEndEmitted)) === 0 && state.length === 0) {
+    state.state |= kEndEmitted;
     stream.emit('end');
 
     if (stream.writable && stream.allowHalfOpen === false) {
```

Comment:
We should maybe move the wState checks below to the bitmap? (with a check so we don't break old writables (e.g. if wState.state maybe?)

Reply:
Future PR? I think that's a bit more risky. We need to move all bimap masks to a util or something.